### ISSUE TITLE
Well defined transformation pipeline for citation identifiers

### ISSUE THREAD

**Issue by dhimmel** [manubot/manubot#114]
We have several types of citation identifiers that potentially required different processing pipelines:

1. citations extracted from a manuscript
2. citations extracted from a manual references file. This category could be subdivided based on what CSL JSON field the citation identifier was extracted from
3. citations passed to the `manubot cite` command

We should think about which processing steps we want to apply in each case. Ideally, we could have modular functions that we pipe together to make it clear and straightforward what transformations will be applied.

Furthermore, we should consider a common nomenclature to use for identifiers based on which transformation steps have been applied. https://github.com/manubot/manubot/pull/113 improves the nomenclature but more is possible.

**Comment by epogrebnyak** [comment]
I got a question: why are you enforcing varieties of identifiers at terminology/notation level and not data structures, eg @dataclass? Terminology convention is a bit hard to follow unless deeply into codebase, while data classes quite unequivocal. 

> Ideally, we could have modular functions

With functions type signatures/hints even better. 

From #129:

> It looks like the major terminology changes are:
> 
>     * `citeproc` to `csl_item`

This is a bibliograhic information in a dict/JSON associated with identifier, right?
 
>     * `citation` and `citation_identifier` to `citekey`

Is this a string starting with `@doi:10/blah` or `doi:10/blah` or `doi, 10/blah` or all of them?

>     * `standard_id` and `standard_citation` to `standard_citekey` (internally)
> 
>     * `short_id` to `short_citekey`
> 

This is point where external reader gets lost. ) Maybe some examples in the docs, or dataclasses even better.

**Comment by epogrebnyak** [comment]
As a follow-up on dataclasses, I think `munubot cite` dataflow can be represented as following sequence of transformations, starting some text string as source (either markdown with `[@citekey]` or one per line citekey text file) and resulting in bibliography rendered by pandoc-cite as text string again. I omit the format (json, palin, etc) for simplicity now.
 
```
text: str -> [CiteKey] -> [Citation] -> [CiteItem] -> csl_style: str -> references: str
```

The main types involved, imo, are `CiteKey`, a tuple of source and indentifier, the `Citation`, parent of `DOI`, `ISBN` and others and ultimately `CiteItem`, a dictionary that is compasable to a CSL JSON.

```
from dataclasses import dataclass

@dataclass
class CiteKey:
    source: str
    identifier: str
    
@dataclass
class Citation:
    identifier: str

    def normalise(self):
        return self 

    def retrieve(self):
        return CiteItem({})

class DOI(Citation):    
    pass
    
class CiteItem(dict):
    def __init__(self, dict_):
        super().__init__(dict_)
        
    def add_id(self):
        return self    
```

Some examples of how these classes might work are [here](https://github.com/epogrebnyak/manubot-cite-minimal/blob/master/explore.py). Any impressions on these data structures, @dhimmel?

**Comment by dhimmel** [comment]
> Any impressions on these data structures

They are interesting and cool to see your progress. I am not sure exactly what problems they solve, or what benefits and disadvantages there would be compared to the current `manubot.cite` implementation.

One disadvantage is that perhaps the functions become slightly less useful for external programs. For example, `DOI('10.1017/aae.2017.13').retrieve()` is perhaps less straightforward to external programs than `citekey_to_csl_item('doi:10.1017/aae.2017.13'). But we could also refactor `citekey_to_csl_item` to call the underlying class-based methods, so the API for this function could stay the same.

Anyways, given that it's a pretty major architectural change that will take considerable effort and possibly introduce (or possibly fix) bugs, I'd want to see clear and major benefits to the approach before proceeding. As far as establishing a well-defined transformation pipeline, I'm not sure a class-architecture is necessary.

**Comment by epogrebnyak** [comment]
Here is how a transformation pipline may look like with classes suggested:

https://github.com/epogrebnyak/manubot-cite-minimal/blob/e49417c95c8f221a49086ac764527f3124e0d3a7/explore.py#L231-L239

**Comment by epogrebnyak** [comment]
> One disadvantage is that perhaps the functions become slightly less useful for external programs. 
For example, DOI('10.1017/aae.2017.13').retrieve() is perhaps less straightforward to external programs than citekey_to_csl_item( )

`DOI('10.1017/aae.2017.13').retrieve()` is a replacement for `get_doi_csl_item(doi)`, not for `citekey_to_csl_item()`. 

You can still have `citekey_to_csl_item()` as a stack of transformations of `CiteKey -> Citation -> CiteItem`.

> But we could also refactor citekey_to_csl_item` to call the underlying class-based methods, so the API for this function could stay the same.

Exactly, here is a simplified version: 

```python
def citekey_to_csl_item(citekey: CiteKey) -> CiteItem:
    return citekey.citation() \
            .normalise() \
            .retrieve() \
            .minimal()
```

On a separate note neither `from manubot.cite import citekey_to_csl_item` nor `from manubot.cite.util import citekey_to_csl_item` can import `citekey_to_csl_item`, results in `ImportError` (maybe circular reference), makes it hard to evaluate usefulness for external programs/scripts. Once made importable, `citekey_to_csl_item()` is certainly a good benchmark for a pipline.

> As far as establishing a well-defined transformation pipeline, I'm not sure a class-architecture is necessary.

I think it is useful to distinguish two sides of 'class-architecture' - one is data structures, another is mounting of methods to classes. One can have `d = DOI('doi:10/blah')` and a `get_doi_csl_item(d: DOI)` function that you already have, this way you separate data structures. You can mount `get_doi_csl_item()` to `DOI` class as `.get_csl_item()` or `.retrieve()` - this way you get the benefit of branching a pipline by class not `if-then` statements (you just call `.get_csl_item()` on any citation item without needing to considr what type it is and the result is guaranteed).

As for the benefits I'd say they are:

- more clear separation of concerns - more visible which methods work on each data structure (not always transparent in existing codebase, enforced by naming convention)
- better testability as result of separation, smaller code under unit test
- less duplication (eg possibly you can have `source, identifier = citekey.split(':', 1)` just once, not several times)
- replacing all the branching like `if source=='doi':`
- you can write all of your pipeline as a sequence of methods as shown above for a minimal `citekey_to_csl_item(citekey: CiteKey) -> CiteItem` example.

Yet given all the reasons above, using dataclasses is a bit of viewpoint shift in python, so it can wait until maintainers are more comfortable with it. In strongly typed languages a pipeline is a sequence of changes from one data structure (type) to another, or operations on these data structures - gives a much safer sense on where is a function or method start and end, make sit easier to reason about the code.

**Comment by dhimmel** [comment]
> On a separate note neither from manubot.cite import citekey_to_csl_item nor from manubot.cite.util import citekey_to_csl_item can import citekey_to_csl_item

Installing the latest version from GitHub should fix this:

```
pip install --upgrade git+https://github.com/manubot/manubot@71d4d3902a343a99c10d59edd6eb77075f9447de
```

> As for the benefits I'd say they are:

Okay, it seems like there could be some benefits, especially relating to consolidating workflow steps and readability. I'd love to see the implementation in terms of a pull request, if that level of contribution interests you. I am on board generally with the idea, but would have to see an implementation so we can be sure it covers the existing functionality.

**Comment by epogrebnyak** [comment]
> I'd love to see the implementation in terms of a pull request

I enjoy getting into the codebase and seeing/changing how things work, but in terms of a single pull request that may be a gigantic change, so it is likely to be rejected. I can work on a mini-fork, ask about implementation detail and use a fork as dataclass-based prototype. 

Given the prototype, existing codebase can be refactored via a series of pull requests, eg starting with `DOI`, `ISBN` and other `Citation` classes, but probably from some other ends - maybe injecting `CiteKey` and `CiteItem` dataclasses first.  `CiteKey -> Citation -> CiteItem` seems a core pipeline as it replicates `citekey_to_csl_item()`, but `CiteKey` and `CiteItem` are less loaded with functionality than `Citation`.  

`Citation` will eliminate some dispatch issues, while `CiteKey` and `CiteItem` are good for readability and delimiting the function scopes boundaries (signatures). Do you have an opinion of what changes you would rather see in your code first? Or all at once?

**Comment by epogrebnyak** [comment]
> Installing the latest version from GitHub should fix this:

That worked, thanks!

**Comment by dhimmel** [comment]
> in terms of a single pull request that may be a gigantic change, so it is likely to be rejected

Good point. It's preferable for pull requests to be focused and small (especially at the beginning). Will make the pace of review and acceptance much faster. If you think starting with a fork that gets merged piece-by-piece is best, then fine with me.

Regarding `CiteKey -> Citation -> CiteItem`, could you provide a bit more description of each class and what it represents?

**Comment by epogrebnyak** [comment]
The fuller chain is `str -> CiteString -> CiteKey -> Citation -> CiteItem`. 

Going from the tail:

- `CiteItem` is a python dictionary with CSL information, widely used as `csl_item` in code
- `Citation` is `DOI()`, `ISBN()` and other source classes. They provide `.csl_item()` method which retrieves `CiteItem`. I think it shoiuld be the only class that interacts with outside world via server calls.
- `CiteKey` is a pair of `source` and `identifier` strings
- `CiteString` just holds `doi:blah` or similar as single string 

There is a bit of temptation to merge `CiteString` and `CiteKey` in one class, but I think there are also benefits to keep them apart. 

```python
@dataclass
class CiteString:
    string: str    

    def citekey(self):
        source, identifier = self.string.split(':', 1)
        return CiteKey(source, identifier)    
    
@dataclass
class CiteKey:    
    source: str
    identifier: str
    
    def citation(self):        
        lookup = dict(doi=DOI, isbn=ISBN) # ~ citeproc_retrievers()
        try:
            constr = lookup[self.source]
        except KeyError:            
            raise ValueError(f'Unsupported source {self.source} in {self}')             
        return constr(self.identifier)

    def __str__(self):
        return f'{self.source}:{self.identifier}'

@dataclass
class Citation:
    identifier: str    
   
    def csl_item(self):
        return CiteItem({})

class CiteItem(dict):
    
    def __init__(self, incoming_dict: dict):
        super().__init__(incoming_dict)
```

An example:

```
CiteString("doi:10/ccg94v")
Out[2]: CiteString(string='doi:10/ccg94v')

CiteString("doi:10/ccg94v").citekey()
Out[3]: CiteKey(source='doi', identifier='10/ccg94v')

CiteString("doi:10/ccg94v").citekey().citation()
Out[4]: DOI(identifier='10/ccg94v')

CiteString("doi:10/ccg94v").citekey().citation().csl_item()
Out[5]: 
{'indexed': {'date-parts': [[2019, 9, 10]],
  'date-time': '2019-09-10T07:43:20Z',
  'timestamp': 1568101400773},
 'reference-count': 0,
 'publisher': 'Springer Nature',
 'issue': '4',
 'content-domain': {'domain': [], 'crossmark-restriction': False},
 'published-print': {'date-parts': [[1990, 4]]},
 'DOI': '10.1038/scientificamerican0490-56',
 'type': 'article-journal',
 'created': {'date-parts': [[2010, 2, 24]],
  'date-time': '2010-02-24T16:16:00Z',
  'timestamp': 1267028160000},
 'page': '56-65',
 'source': 'Crossref',
 'is-referenced-by-count': 355,
 'title': 'The Unusual Origin of the Polymerase Chain Reaction',
 'prefix': '10.1038',
 'volume': '262',
 'author': [{'given': 'Kary B.',
   'family': 'Mullis',
   'sequence': 'first',
   'affiliation': []}],
 'member': '297',
 'container-title': 'Scientific American',
 'original-title': [],
 'deposited': {'date-parts': [[2011, 8, 9]],
  'date-time': '2011-08-09T13:40:17Z',
  'timestamp': 1312897217000},
 'score': 1.0,
 'subtitle': [],
 'short-title': [],
 'issued': {'date-parts': [[1990, 4]]},
 'references-count': 0,
 'journal-issue': {'published-print': {'date-parts': [[1990, 4]]},
  'issue': '4'},
 'alternative-id': ['scientificamerican0490-56'],
 'URL': 'https://doi.org/10/ccg94v',
 'relation': {},
 'ISSN': ['0036-8733'],
 'subject': ['Multidisciplinary'],
 'container-title-short': 'Sci Am'}

ci = CiteString("doi:10/ccg94v").citekey().citation().csl_item()

ci.title
Out[7]: 'The Unusual Origin of the Polymerase Chain Reaction'

ci.author[0]['family']
Out[8]:  'Mullis'
```

Finally, a key interface function `citekey_to_csl_item` may look something like this (disregarding logging):

```        
def citekey_to_csl_item(citekey_str: str, prune=True) -> dict:
    """
    Generate a CSL item (Python dictionary) for the input *citekey* string.
    """
    csl_item = CiteString(citekey_str).citekey().citation().standardize().csl_item()
    csl_item.fix_type()
    csl_item.add_note()
    csl_item.generate_id()
    if prune:
        csl_item.prune()
    return csl_item
```

> ... starting with a fork that gets merged piece-by-piece is best

Maybe working through a test suit (e.g. splitting end-to-end tests for whole API and unit test for parts of functionality) and original code refactoring before introducing the dataclasses can help. It is a tricky point how to migrate the codebase incrementally though, need to think it over. 

Have you thought of isolated repo like `manubot/manubot-cite` or  `manubot/manubot-cite` for this part of the project - that would make transition a bit easier, at least less thoughts about relations to other part of the project, also easier to focus on the changes.

**Comment by agitter** [comment]
As we're discussing this major restructuring with the more object-oriented `dataclass` workflow, I'd also like to consider the transformation pipeline for references.  It could be a good idea to unify some of the classes or transformations that are applied to citations and references.  The distinction between the two is partly defined in https://github.com/manubot/manubot/pull/129#discussion_r321985026

**Comment by epogrebnyak** [comment]
> Now that we are refining the terminology, it may be helpful to create a glossary.

@agitter - I noticed a mention of a glossary at https://github.com/manubot/manubot/pull/129#pullrequestreview-282304372, it would be very useful.  

In our best case the classes would fit the mind model and a glossary.

**Comment by epogrebnyak** [comment]
Got more human-readable version of the classes:

```
    CiteKey -> Handle -> CSL_Dict
    CiteKey('doi:10/gf8gkw') -> DOI('10.1038/171737a0') -> {'title': ...}
```

`CiteKey` holds the source-prefixed string with identifier, `Handle` is parent of `DOI` and fiends and `CSL_Dict` is a dictionary with extra methods to clean it. 

```python
ci = CiteKey('doi:10.1038/171737a0').handle().csl_dict()

assert ci['title'].startswith('Molecular Structure of Nucleic Acids')
assert standardize_citekey('doi:10/gf8gkw') == 'doi:10.1038/d41586-019-02781-4'
assert citekey_to_csl_item('doi:10/gf8gkw')['title'] == \
    'How to make computing more sustainable'
```

Draft implementation: https://github.com/epogrebnyak/manubot-cite-minimal/blob/23e28d3a91f486aac90d41ff3417bb1a9781f9c8/citekey.py

**Comment by dhimmel** [comment]
@epogrebnyak nice progress on your implementation. Very readable and self-documenting.

I'm curious to see how this will eventually interact with `manubot process`. See for example:

https://github.com/manubot/manubot/blob/71d4d3902a343a99c10d59edd6eb77075f9447de/manubot/process/util.py#L205-L243

I'll be offline a fair amount in the coming week, but seems like you've got a good grip on how to proceed.

**Comment by epogrebnyak** [comment]
Thanks for pointing out to get_citekeys_df(), helps on terminology and pipeline level, let me some time to digest 

As of now we are not breaking any 'manubot cite' functions, just changing some implementation, so get_citekeys_df() should not get hurt.

My plan is the following:
- experiment with new classes in manubot-cite-minimal
- use existing function imports from manubot to run this implementation
- do PRs on splitting some functions to smaller ones so that they correspond to individual methods (specifically CSL_Dict)
- document corresponding changes that are needed to inject dataclasses to manubot cite

After that the project will be more ready to introducing classes

**Comment by epogrebnyak** [comment]
- manuscript_citekey: citation keys extracted from the manuscript content files. 
	     - detagged_citekey: manuscript_citekey but with tag citekeys dereferenced 
	     - standard_citekey: detagged_citekey standardized 
	     - short_citekey: standard_citekey hashed to create a shortened citekey 

All these are part of transformation, thinking of incorporating into dataclasses and their methods.
